### PR TITLE
Add --dangling option to rmi command

### PR DESCRIPTION
**Pull request by artmello**
Hi,

After using buildah for a while I ended up with many dangling images at my images list, as described at issue #410. Checking the comments on such issue it was suggested adding a --dangling option to rmi. This is my try to implement such a feature. If you think this is not the correct approach, I am happy to re-work this patch. 

Thanks

### CODE REVIEW

**Comment by TomSweeneyRedHat** [comment]
bot, add author to whitelist

**Comment by TomSweeneyRedHat** [comment]
@artmello Thanks for the submit!  It doesn't look like you signed your commit.  If you could do 'git commit --amend -s' and then repush the submit, that would be great.  Also an added test or two to tests/rmi.bats would be awesome.

The code that you have looks good to me on the surface.  However Travis is failing with:

vendor/github.com/ostreedev/ostree-go/pkg/otremote/remoteadd.go:1:1: expected 'package', found 'EOF'
cmd/buildah/rmi.go:43::warning: cyclomatic complexity 41 of function rmiCmd() is high (> 40) (gocyclo)
make: *** [validate] Error 1

@rhatdan @nalind is this something we can override, or do we have to rework the code to get the number of if statements/checks down in that function?

**Comment by rhatdan** [comment]
@artmello Make your code into a function removeDangling() and call it, that should fix the complexity issue.

**Comment by artmello** [comment]
Thanks everyone for reviewing. I changed the code so hopefully complexity under rmiCmd() is reduced and commit should be signed. Still missing some new tests for it, that I will be working next.

**Comment by rhatdan** [comment]
Can you add a test.

**Comment by artmello** [comment]
New test added

**Comment by rhatdan** [comment]
@rh-atomic-bot r+

**Comment by TomSweeneyRedHat** [comment]
The changes LGTM, but after to testing/thinking about this, I'd like to change the name of the parameter from "dangling" to "prune".  Two reasons.  I'm not at  all happy with the short version -d as it's generally used for delete.  Secondly the functionality that does this in Docker is 'docker system prune'.  

@rhatdan WDYT?  I see this is close to being merged.  I'm happy to change it afterwards.

**Comment by artmello** [comment]
On those lines, would make sense instead of having --dangling option, having a --filter one as we have for "images" command. Would be a little more verbose than --dangling (would be something like --filter "dangling=true"), but you could remove all images that meet some filter and not dangling ones. WDYT?

**Comment by TomSweeneyRedHat** [comment]
Personally I'd rather the prune than the filter.  Part of that is personal bias, I've never been a big fan of filter options (yet I grep output like crazy), and I'd like to somewhat match the functions that Docker provides.  @rhatdan and @nalind are the folks that set the valid Buildah commands, I'll let them weigh in.

**Comment by rhatdan** [comment]
I like prune.

**Comment by artmello** [comment]
I substitute --dangling,-d option to --prune,-p on rmi command. WDYT?

**Comment by artmello** [comment]
@TomSweeneyRedHat @rhatdan I saw CI failed. Should I close PR and reopen it?

File: docs/buildah-rmi.md [hunk 1]
```diff
@@ -27,6 +31,8 @@ buildah rmi --all
 
 buildah rmi --all --force
 
+buildah rmi --dangling
```
Comment (TomSweeneyRedHat) [inline docs/buildah-rmi.md]: prune

File: cmd/buildah/rmi.go [hunk 2]
```diff
@@ -21,6 +21,10 @@ var (
 			Name:  "all, a",
 			Usage: "remove all images",
 		},
+		cli.BoolFlag{
+			Name:  "prune, p",
+			Usage: "remove all dangling images",
```
Comment (TomSweeneyRedHat) [inline cmd/buildah/rmi.go]: "remove all" to 'prune"

File: cmd/buildah/rmi.go [hunk 3]
```diff
@@ -39,14 +43,18 @@ var (
 func rmiCmd(c *cli.Context) error {
 	force := c.Bool("force")
 	removeAll := c.Bool("all")
+	removeDangling := c.Bool("prune")
```
Comment (TomSweeneyRedHat) [inline cmd/buildah/rmi.go]: Just to be consistent and to make it easier to find, can you change removeDangling to pruneDangling throughout?  I'll try to mark them.

File: cmd/buildah/rmi.go [hunk 4]
```diff
@@ -39,14 +43,18 @@ var (
...
 
 	args := c.Args()
-	if len(args) == 0 && !removeAll {
+	if len(args) == 0 && !removeAll && !removeDangling {
```
Comment (TomSweeneyRedHat) [inline cmd/buildah/rmi.go]: pruneDangling

File: cmd/buildah/rmi.go [hunk 5]
```diff
@@ -39,14 +43,18 @@ var (
...
 		return errors.Errorf("image name or ID must be specified")
 	}
 	if len(args) > 0 && removeAll {
 		return errors.Errorf("when using the --all switch, you may not pass any images names or IDs")
 	}
+	if removeAll && removeDangling {
```
Comment (TomSweeneyRedHat) [inline cmd/buildah/rmi.go]: pruneDangling

File: cmd/buildah/rmi.go [hunk 6]
```diff
@@ -61,12 +69,16 @@ func rmiCmd(c *cli.Context) error {
...
+		imagesToDelete, err = removeAllImages(store)
 		if err != nil {
-			return errors.Wrapf(err, "error reading images")
+			return err
 		}
-		for _, image := range images {
-			imagesToDelete = append(imagesToDelete, image.ID)
+	}
+
+	if removeDangling {
```
Comment (TomSweeneyRedHat) [inline cmd/buildah/rmi.go]: pruneDangling

File: cmd/buildah/rmi.go [hunk 7]
```diff
@@ -61,12 +69,16 @@ func rmiCmd(c *cli.Context) error {
...
+		imagesToDelete, err = removeDanglingImages(store)
```
Comment (TomSweeneyRedHat) [inline cmd/buildah/rmi.go]: pruneDanglingImages

File: cmd/buildah/rmi.go [hunk 8]
```diff
@@ -272,3 +284,35 @@ func storageImageID(store storage.Store, id string) (types.ImageReference, error
 	}
 	return nil, errors.Wrapf(err, "error parsing %q as a storage image reference: %v", id)
 }
+
+// Returns a list of all existing images to be removed
+func removeAllImages(store storage.Store) ([]string, error) {
```
Comment (TomSweeneyRedHat) [inline cmd/buildah/rmi.go]: You're not really removing anything in this function, could we rename it to findAllImages?

File: cmd/buildah/rmi.go [hunk 9]
```diff
@@ -272,3 +284,35 @@ func storageImageID(store storage.Store, id string) (types.ImageReference, error
...
+	}
+	for _, image := range images {
+		imagesToDelete = append(imagesToDelete, image.ID)
+	}
+
+	return imagesToDelete, nil
+}
+
+// Returns a list of all dangling images to be removed
+func removeDanglingImages(store storage.Store) ([]string, error) {
```
Comment (TomSweeneyRedHat) [inline cmd/buildah/rmi.go]: ditto, rename to findDanglingImages

**Comment by TomSweeneyRedHat** [comment]
@artmello really good work on this.  I just have a bunch of minor nits to tend to that will make long term code support more clear.  Thank you!  Plus hopefully a new submit will cleanse the Red Hat CI daemon issues away.  I'm sure it's not related at all to your changes.

**Comment by TomSweeneyRedHat** [comment]
Sorry, hit the wrong button, didn't mean to close it.

**Comment by artmello** [comment]
@TomSweeneyRedHat thanks a lot for your review. I tried to address the comments on this version.

**Comment by rhatdan** [comment]
@rh-atomic-bot r+

**Pull request closed without merging**